### PR TITLE
update how variable augmentations work

### PR DESCRIPTION
**Pull request by jakemac53**
Closes https://github.com/dart-lang/language/issues/3046.
Partly resolves https://github.com/dart-lang/language/issues/3041.

I altered the proposal to simply disallow augmenting getters and setters with variables. While this is potentially possible to specify, I think it is difficult to explain the behavior and wouldn't be intuitive. We can always add it back later.

In particular, I think we may later add the ability to augment abstract/external getters/setters with variables.

@johnniwinther do you know if this would cause any problems for the patch file -> augmentation library conversions? Do we implement any getters and/or setters with fields?

### CODE REVIEW

**Review by johnniwinther — APPROVED** [review]
Patching is currently not supported for fields so this change is not in conflict with implementing patches as augmentations.

File: working/augmentation-libraries/feature-specification.md [hunk 1]
```diff
@@ -518,14 +530,14 @@ More specifically:
...
-
 It is a compile-time error if:
 
 *   The original and augmenting declarations do not have the same type.
 
 *   An augmenting declaration uses `augment super` when the original declaration
-    is marked `external`.
+    has no concrete implementation. It **is** allowed to call `augment super`
+    when augmenting an external declaration _if_ an implementation has been
+    filled in by a compiler or other external source.
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:540]: That if is platform/implementation-specific. It's probably better to just say that unconditionally that all `external` declarations are treated as having implementations.

File: working/augmentation-libraries/feature-specification.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Augmentation Libraries
 
 Author: rnystrom@google.com, jakemac@google.com
-Version: 1.10 (see [Changelog](#Changelog) at end)
+Version: 1.11 (see [Changelog](#Changelog) at end)
```
Comment (munificent) [inline working/augmentation-libraries/feature-specification.md:4]: We also talked about extending the language to allow abstract member declarations inside non-abstract classes (providing that they get augmented with a body before the class is compiled).

I'm not sure where that should go in here, but probably somewhere.

**Review by munificent — APPROVED** [review]

File: working/augmentation-libraries/feature-specification.md [hunk 3]
```diff
@@ -492,24 +489,39 @@ More specifically:
...
+    augmenting abstract or external getters and setters with variables, but for
+    now you can instead use the following workaround:
+
+    - Add a new field.
+    - Augment the getter and/or setter to delegate to that field.
+
+    If a concrete variable is augmented by a getter or setter, you **can** still
+    augment the variable, as you are only augmenting the initializer. This is
+    not considered to be augmenting the augmenting getter or setter, since those
+    are not actually altered.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:503]: I took that mostly word for word with one additional sentence

File: working/augmentation-libraries/feature-specification.md [hunk 4]
```diff
@@ -492,24 +489,39 @@ More specifically:
...
+    If the variable being augmented **does not** have a concrete implementation,
+    then it gets one from the augmenting variable. This includes the backing
+    store, the implicit getter and setter, as well as the initializer if
+    present.
+
+    If the variable being augmented **does** have a concrete implementation,
+    then only the initializer has any meaning, and it replaces the original
+    initializer. In this case the augmenting initializer may use an
     `augment super` expression which executes the original initializer
     expression when evaluated.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:520]: Worded a bit differently but done

File: working/augmentation-libraries/feature-specification.md (hunk 2 above)
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md]: I did that in one of the other PRs :)

**Pull request merged**